lib/rubygems/ext/cmake_builder.rb

```diff
@@ -1,21 +1,114 @@
 # frozen_string_literal: true
 
+# This builder creates extensions defined using CMake. Its is invoked if a Gem's spec file
+# sets the `extension` property to a string that contains `CMakeLists.txt`.
+#
+# In general, CMake projects are built in two steps:
+#
+#  * configure
+#  * build
+#
+# The builder follow this convention. First it runs a configuration step and then it runs a build step.
+#
+# CMake projects can be quite configurable - it is likely you will want to specify options when
+# installing a gem. To pass options to CMake specify them after `--` in the gem install command. For example:
+#
+#   gem install <gem_name> -- --preset <preset_name>
+#
+# Note that options are ONLY sent to the configure step - it is not currently possible to specify
+# options for the build step. If this becomes and issue then the CMake builder can be updated to
+# support build options.
+#
+# Useful options to know are:
+#
+#  -G to specify a generator (-G Ninja is recommended)
+#  -D<CMAKE_VARIABLE> to set a CMake variable (for example -DCMAKE_BUILD_TYPE=Release)
+#  --preset <preset_name> to use a preset
+#
+# If the Gem author provides presets, via CMakePresets.json file, you will likely want to us use one of them.
+# If not, then you may wish to specify a generator (Ninja is recommended since it can build projects in parallel
+# versus a Make generator which build ins serial and thus is *much* slower).
+
+require "fileutils"
+
 class Gem::Ext::CmakeBuilder < Gem::Ext::Builder
-  def self.build(extension, dest_path, results, args=[], lib_dir=nil, cmake_dir=Dir.pwd,
-    target_rbconfig=Gem.target_rbconfig)
+  attr_accessor :runner, :profile
+
+  def initialize
+    @runner = self.class.method(:run)
+    @profile = :release
+  end
+
+  def build(extension, gem_dir, results, args = [], lib_dir = nil, cmake_dir = Dir.pwd, target_rbconfig = Gem.target_rbconfig)
     if target_rbconfig.path
       warn "--target-rbconfig is not yet supported for CMake extensions. Ignoring"
     end
 
-    unless File.exist?(File.join(cmake_dir, "Makefile"))
-      require_relative "../command"
-      cmd = ["cmake", ".", "-DCMAKE_INSTALL_PREFIX=#{dest_path}", *Gem::Command.build_args]
+    # Make sure lib dir is set
+    lib_dir ||= File.join(gem_dir, "lib")
```

Comment:
I don't know why `lib_dir` is optional, since RubyGems always passes it when building extensions. We could consider making it non optional but for now I'd go with ignoring anything that depends on `lib_dir` (I guess it's just passing `-DCMAKE_RUNTIME_OUTPUT_DIRECTORY` and friends?), rather that giving it a default value. Again, for consistency with other builders.

Reply:
I did that because when running tests it was not always set and that caused the test to fail. See https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_ext_cmake_builder.rb#L42.

I will update the test instead.